src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs

```diff
@@ -75,5 +144,96 @@ public PropertyBuilder DefineProperty(string name, PropertyAttributes attributes
                 => DefineProperty(name, attributes, default,
                     returnType, returnTypeRequiredCustomModifiers, returnTypeOptionalCustomModifiers,
                     parameterTypes, parameterTypeRequiredCustomModifiers, parameterTypeOptionalCustomModifiers);
+
+        public virtual PropertyBuilder DefineProperty(string name, PropertyAttributes attributes, CallingConventions callingConvention,
+            Type returnType, Type[]? returnTypeRequiredCustomModifiers, Type[]? returnTypeOptionalCustomModifiers,
+            Type[]? parameterTypes, Type[][]? parameterTypeRequiredCustomModifiers, Type[][]? parameterTypeOptionalCustomModifiers)
+                => DefineProperty(name, attributes, callingConvention,
+                    returnType, returnTypeRequiredCustomModifiers, returnTypeOptionalCustomModifiers,
+                    parameterTypes, parameterTypeRequiredCustomModifiers, parameterTypeOptionalCustomModifiers);
+
+        public virtual ConstructorBuilder DefineTypeInitializer()
+            => DefineTypeInitializer();
+
+        public virtual FieldBuilder DefineUninitializedData(string name, int size, FieldAttributes attributes)
+            => DefineUninitializedData(name, size, attributes);
+
+        public static ConstructorInfo GetConstructor(Type type, ConstructorInfo constructor)
+        {
+            Module module = type.Module;
+            if (module is not ModuleBuilder moduleBuilder)
+            {
+                if (type.IsGenericType)
+                {
+                    foreach (Type t in type.GetGenericArguments())
+                    {
+                        if (t.Module is ModuleBuilder mb)
+                        {
+                            return mb.GetConstructor(type, constructor);
+                        }
+                    }
+                }
+
+                throw new ArgumentException(SR.Argument_MustBeTypeBuilder, nameof(type));
```

Comment:
This is potential breaking change as we have discussed on Teams.

The following code works fine today, but it is going to fail after this change:
```csharp
using System.Reflection;
using System.Reflection.Emit;

var nonRuntimeInstantiation = typeof(List<>).MakeGenericType(new TypeDelegator(typeof(int)));
var ctor = typeof(List<>).GetConstructor(Type.EmptyTypes)!;
Console.WriteLine(TypeBuilder.GetConstructor(nonRuntimeInstantiation, ctor));
```

Reply:
> This is potential breaking change as we have discussed on Teams.

Right, it is not final solution, I just wanted to see if any other failures show up

> The following code works fine today, but it is going to fail after this change:

Though I am in doubt if we still should keep supporting it. If I understood correctly this approach is not by design and were not supposed work, it seems in the above case [the mono runtime might throw](https://github.com/dotnet/runtime/blob/main/src/mono/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.Mono.cs#L1759-L1780). If it was not by design, I would assume using such `ConstructorInfo` would not work properly for all cases, though I might be wrong.

Reply:
> it seems in the above case [the mono runtime might throw](https://github.com/dotnet/runtime/blob/main/src/mono/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.Mono.cs#L1759-L1780)

I do not see where it would throw for this case. If I am reading the code correctly. `type` is `TypeBuilderInstantiation` in this case, so the very first check `if (type is TypeBuilder || type is TypeBuilderInstantiation)` is going to succeed.

Mono seems to be actually more permissive compared to CoreCLR in this case. CoreCLR requires the `type` to be TypeBuilder or TypeBuilderInstantiation, and throws otherwise. Mono seems to allow more System.Type variants here.

Reply:
> type is TypeBuilderInstantiation in this case

Right, was not aware of that

Reply:
Update: we decided to keep the old behavior